typing/typecore.ml

```diff
@@ -3140,8 +3159,8 @@ and type_expect_
             and (cty', force') =
               Typetexp.transl_simple_type_delayed env sty'
             in
-            let ty = cty.ctyp_type in
-            let ty' = cty'.ctyp_type in
+            let ty = instance cty.ctyp_type in
+            let ty' = instance cty'.ctyp_type in
```

Comment:
There remains occurrences of `.ctyp_type` that are not protected by `instance` after the PR:
- line 2402 in `create_package_type`
- line 3000 in a `Pexp_constraint` case
- line 3477 in a `Pexp_poly` type

If I understand correctly, the assumption is that "cty"s obtained from `transl_simple_type_delayed` must be `instance`d, but those by `transl_simple_type` need not.

Given that every single call to one of those two functions in the module is followed by a projection of the `ctyp_type` field, wouldn't it make sense to have the function itself return this value? Then the `_delayed` version would be careful to call `instance` itself, and no special discipline would be required on the consumer side.

Reply:
> If I understand correctly, the assumption is that "cty"s obtained from `transl_simple_type_delayed` must be `instance`d, but those by `transl_simple_type` need not.

Correct.

> Given that every single call to one of those two functions in the module is followed by a projection of the `ctyp_type` field, wouldn't it make sense to have the function itself return this value?

No. While indeed we do take an instance straightaway (sometime to unify with the expected type, sometime to pass forward as an expected type, etc), the value itself with its *generalized* type is being stored in the typedtree.
And that generic type that is being stored, is being reused later on in `build_as_type`, to get a fresh instance (read: minimal sharing) of the constraint.

That is also why we don't want to do this:
> Then the `_delayed` version would be careful to call instance itself, and no special discipline would be required on the consumer side.

---

To view it in terms of your rigid/flexible proposition: the "generic" type serves the same role as your "rigid" one. And each instance of it, is a fresh "flexible" one.

Reply:
I was merely suggesting that `transl_type_simple` could return a triple `(ty, cty, force)` with `ty = cty.ctyp_type`, and `transl_type_delayed` a triple `(ty, cty, force)` with `ty = instance cty.ctyp_type`. I have the impression that your reply is assuming that I propose to store the one-time instance in the `cty` value.

Reply:
Ah, indeed I had misunderstood your proposition.

Reply:
Done.